iotilesensorgraph/iotile/sg/graph.py

```diff
@@ -317,6 +318,45 @@ def get_tick(self, name):
         except ArgumentError:
             return 0
 
+    def add_checksum(self):
+        """Check metadata if sensorgraph's checksum needs to be added."""
+
+        if 'hash_address' not in self.metadata_database or\
+            'hash_algorithm' not in self.metadata_database:
+            return
+
+        slot = SlotIdentifier.FromString('controller')
+        config_id = self.metadata_database['hash_address']
+        algorithm = self.metadata_database['hash_algorithm']
+
+        hash_algorithm = KNOWN_HASH_ALGORITHMS[algorithm]
+
+        nodes_checksum = hash_algorithm.calculate(hash_algorithm.algorithm,
+                                                  self.get_nodes_binary())
+        self._logger.debug("nodes_checksum: %s", nodes_checksum)
+        streamers_checksum = hash_algorithm.calculate(hash_algorithm.algorithm,
+                                                  self.get_streamers_binary())
+        self._logger.debug("streamers_checksum: %s", streamers_checksum)
+        configs_checksum = hash_algorithm.calculate(hash_algorithm.algorithm,
+                                                  self.get_config_database_binary(ignore_configs=[config_id]))
+        self._logger.debug("configs_checksum: %s", configs_checksum)
+        constants_checksum = hash_algorithm.calculate(hash_algorithm.algorithm,
+                                                  self.get_constant_database_binary())
+        self._logger.debug("constants_checksum: %s", constants_checksum)
+        metadata_checksum = hash_algorithm.calculate(hash_algorithm.algorithm,
+                                                  self.get_metadata_database_binary())
+        self._logger.debug("metadata_checksum: %s", metadata_checksum)
+        combined_checksum_string = nodes_checksum + streamers_checksum +\
+                                   configs_checksum + constants_checksum +\
+                                   metadata_checksum
+        combined_checksum_bytes = bytes(combined_checksum_string, 'utf-8')
+        device_checksum = hash_algorithm.calculate(hash_algorithm.algorithm, 
+                                                   combined_checksum_bytes)
+        self._logger.debug("device_checksum: %s", device_checksum)
+
+        config_type, _ = self.get_config(slot, config_id)
+        self.add_config(slot, config_id, config_type, device_checksum)
```

Comment:
This replaces the value of the `config_id` variable, which the user specifies as the address that matches the `meta hash_address` value.